qiskit/algorithms/eigensolvers/vqd.py

```diff
@@ -0,0 +1,431 @@
+                    self.estimator, self.ansatz, aux_operators, result.optimal_point[-1]
+                )
+                aux_values.append(aux_value)
+
+            if step == 1:
+
+                logger.info(
+                    "Ground state optimization complete in %s seconds.\n"
+                    "Found opt_params %s in %s evals",
+                    eval_time,
+                    result.optimal_point,
+                    self._eval_count,
+                )
+            else:
+                logger.info(
+                    (
+                        "%s excited state optimization complete in %s s.\n"
+                        "Found opt_params %s in %s evals"
+                    ),
+                    str(step - 1),
+                    eval_time,
+                    result.optimal_point,
+                    self._eval_count,
+                )
+
+        # To match the signature of EigensolverResult
+        result.eigenvalues = np.array(result.eigenvalues)
+        result.optimal_point = np.array(result.optimal_point)
+        result.optimal_value = np.array(result.optimal_value)
+        result.cost_function_evals = np.array(result.cost_function_evals)
+        result.optimizer_time = np.array(result.optimizer_time)
+
+        if aux_operators is not None:
+            result.aux_operator_eigenvalues = aux_values
+
+        return result
+
+    def _get_evaluate_energy(
+        self,
+        step: int,
+        operator: BaseOperator | PauliSumOp,
+        prev_states: list[np.ndarray] | None = None,
+    ) -> Callable[[np.ndarray], float | list[float]]:
+        """Returns a function handle to evaluate the ansatz's energy for any given parameters.
+            This is the objective function to be passed to the optimizer that is used for evaluation.
+
+        Args:
+            step: level of energy being calculated. 0 for ground, 1 for first excited state...
+            operator: The operator whose energy to evaluate.
+            prev_states: List of parameters from previous rounds of optimization.
+
+        Returns:
+            A callable that computes and returns the energy of the hamiltonian
+            of each parameter.
+
+        Raises:
+            RuntimeError: If the circuit is not parameterized (i.e. has 0 free parameters).
+            AlgorithmError: If operator was not provided.
+
+        """
+
+        num_parameters = self.ansatz.num_parameters
+        if num_parameters == 0:
+            raise RuntimeError("The ansatz must be parameterized, but has 0 free parameters.")
+
+        if step > 1 and (len(prev_states) + 1) != step:
+            raise RuntimeError(
+                f"Passed previous states of the wrong size."
+                f"Passed array has length {str(len(prev_states))}"
+            )
+
+        self._check_operator_ansatz(operator)
+
+        prev_circs = []
+        for state in range(step - 1):
+            prev_circs.append(self.ansatz.bind_parameters(prev_states[state]))
```

Comment:
I imagine we could build up these circuits just once as we go step by step. Here it will be rebuilding (binding) them each time which will be seen as different circuits on down through the primitives. Maybe this can be some future improvement in the interests of time if its not easily changed.

Reply:
Will leave as a future improvement. Not resolving the conversation for that purpose.